cpp/src/arrow/engine/substrait/expression_internal.cc

```diff
@@ -157,14 +157,15 @@ Result<compute::Expression> FromProto(const substrait::Expression& expr,
     case substrait::Expression::kScalarFunction: {
       const auto& scalar_fn = expr.scalar_function();
 
-      auto id = ext_set.function_ids()[scalar_fn.function_reference()];
+      ARROW_ASSIGN_OR_RAISE(auto decoded_function,
+                            ext_set.DecodeFunction(scalar_fn.function_reference()));
 
       std::vector<compute::Expression> arguments(scalar_fn.args_size());
       for (int i = 0; i < scalar_fn.args_size(); ++i) {
         ARROW_ASSIGN_OR_RAISE(arguments[i], FromProto(scalar_fn.args(i), ext_set));
       }
 
-      return compute::call(id.name.to_string(), std::move(arguments));
+      return compute::call(decoded_function.name.to_string(), std::move(arguments));
```

Comment:
This is actually a functionality change.  I think the old behavior was incorrect.  When we make an Arrow `call` expression we should be using the Arrow name and not the Substrait name.  This wasn't caught in tests because the only extension function used by the tests (add) has the same value for both fields.

If you don't take the simplification then still make sure you include this change.

Reply:
ARROW-15583

Reply:
Check, it looks like `function_names()` was intended to be used for this, but was never actually used at all (well, it was once, in a test).